### ISSUE TITLE
ML module milestones

### ISSUE THREAD

**Issue by laksh22** [scorelab/LabelLab#425]
Detailed milestones for Issue #285 
> Note: [Decisions taken during community bonding period](https://docs.google.com/document/d/1YuMbvlFgim31wdXnmMAKDJ1w1YuqVvv3PMdUuDCM_4k/edit?usp=sharing)

<details>
  <summary>See UI mockups</summary>

__Model management dashboard__
![model_manager](https://user-images.githubusercontent.com/9462834/82452443-d7f99a80-9ae1-11ea-9ce3-7f59dd11e216.PNG)

__New model options__
![new_model](https://user-images.githubusercontent.com/9462834/82452486-e6e04d00-9ae1-11ea-9489-cdbe5d1b2165.PNG)

__Model testing__
![test_model](https://user-images.githubusercontent.com/9462834/82452506-eba50100-9ae1-11ea-9978-828235f801e7.PNG)

__Custom classifier editor__
![custom_model](https://user-images.githubusercontent.com/9462834/82452528-f2cc0f00-9ae1-11ea-8eb8-4eba6b87ac2a.PNG)

__Transfer learning(Keras API/saved model) classifier editor__
![t_learning](https://user-images.githubusercontent.com/9462834/82452563-ffe8fe00-9ae1-11ea-9563-7619301cc88d.PNG)

__Transfer learning(uploaded model) classifier editor__
![uploaded_model](https://user-images.githubusercontent.com/9462834/82452570-024b5800-9ae2-11ea-9608-9f65ded8ae62.PNG)

</details>

## Milestone 1  
- [x] Create UI of model management dashboard
- [x] Create basic layout for the ML module UI
- [x] Create initial Redux store/actions/reducers for the model
- [x] Create SQL Table structure for model
- [x] Create module to add training class to the model
    - [x] Create UI
    - [x] Integrate with Redux
- [x] Create module to choose image pre-processing steps
    - [x] Create UI
    - [x] Integrate with Redux
- [x] Create module to perform transfer learning on a Keras Application
    - [x] Create UI
    - [x] Build a parser in the backend to parse model details stored in database into a `tf.keras.Model`
    - [x] Create backend
    - [x] Integrate with backend
    - [x] Add needed dependencies for machine learning (tensorflow, etc.)
- [x] Add functionality to save changes made to a model
    - [x] Create UI
    - [x] Create backend
    - [x] Integrate with backend
- [x] Create module to show accuracy and loss graphs after training
    - [x] Create UI
    - [x] Create backend
    - [x] Integrate with back-end

## Milestone 2
- [x] Create module to build custom model (add/edit/remove layers)
    - [x] Create UI
    - [x] Integrate with Redux
    - [x] Create backend
- [x] Create module for exporting a trained model in 3 formats
    - [x] Create UI
    - [x] Integrate with Redux
    - [x] Create backend
        - [x] Keras .h5
        - [x] SavedModel
        - [x] ONNX
- [x] Create module for testing a trained model (inside the model builder dashboard)
    - [x] Create UI
    - [x] Integrate with Redux
    - [x] Create backend

## Milestone 3
- [x] Create module for uploading a model for training
    - [x] Create UI
    - [x] Integrate with Redux
    - [x] Create backend
- [x] Modify transfer learning module to also choose and train saved models
    - [x] Modify UI
    - [x] Integrate with Redux
    - [x] Create backend
- [x] Add model testing to the model management dashboard
   - [x] Create UI
   - [x] Integrate with Redux
   - [x] Create backend
- [x] Integrate model deletion/editing with the model management dashboard
    - [x] Create UI
    - [x] Integrate with Redux
    - [x] Create backend

**Comment by laksh22** [comment]
> Many-many between model and class
> Many-one between preprocessing step and model
> Many-one between layers and model

__Model table__
```sql
id INT AUTO_INCREMENT,
name VARCHAR,
type VARCHAR, -- classifier, segmentation, object detection, etc.
train_split DECIMAL,
test_split DECIMAL,
validation_split DECIMAL,
epochs INT,
batch_size INT,
learning_rate INT,
loss_type VARCHAR,
optimizer: VARCHAR,
metrics: VARCHAR,
loss_graph_url: VARCHAR, -- URL to access the graph of the training loss
accuracy_graph_url: VARCHAR, -- URL to access the graph of the training accuracy
saved_model_url: VARCHAR -- URL to access the saved model if the model has been trained
```

__Model-Class junction table__
```sql
class_id INT,
model_id INT
```

__Preprocessing steps table__
```sql
attributes?,
model_id INT
```
Not sure how to create this table(attributes section) since the number and names of the fields will depend on the type of preprocessing step chosen and the options available for that step.

__Layers table__
```sql
attributes?,
model_id INT
```

Similarly to the preprocessing table, the number and names of the fields depends on the layer chosen and the options available for that layer.

**Comment by carpecodeum** [comment]
Hey @laksh22 I'm not sure about what you exactly want help in working with SQL I hope that you can elaborate that a bit. Also, inorder to help you I can recommend going through [this](https://flask-sqlalchemy.palletsprojects.com/en/2.x/) I will be making a PR most probably tomorrow with the models and maybe you can refer that :)

**Comment by cazci** [comment]
Hey, shall we design an ER for a better clearance? It'll be helpful for future maintenance as well.

**Comment by ivantha** [comment]
> Hey, shall we design an ER for a better clearance? It'll be helpful for future maintenance as well.

+1. 
Please go with MySQL Workbench.

**Comment by laksh22** [comment]
@carpecodeum My query was regarding the `layers` and `preprocessing_steps` tables. Each row in the layers table for example should contain the name of a layer of a sequential model as well as its attributes. For example, a layer of the type `Conv2D` will have the fields `filters`, `kernel_size`, `activation` while a layer of the type `Dense` will just have a field named `units`. What is an ideal way to store such data where each row can have a different number of attributes? 

Also thanks for the link, I'll check it out!

**Comment by carpecodeum** [comment]
@laksh22 I think @UdeshUK @ivantha @Tushar19varshney Might be able to help with that better :)
